tests/integration_tests/test_public_api.py

```diff
@@ -0,0 +1,144 @@
+import ipaddress
+from contextlib import contextmanager, suppress
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Generator, Type
+
+import pytest
+
+import pycloudlib
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.instance import BaseInstance
+from pycloudlib.util import LTS_RELEASES
+
+cloud_config = """\
+#cloud-config
+runcmd:
+  - echo 'hello' > /var/tmp/example.txt
+"""
+
+
+@pytest.fixture
+def cloud(request):
+    cloud_instance: BaseCloud = request.param(
+        tag="pycloudlib-integration",
+        timestamp_suffix=True,
+    )
+
+    if isinstance(cloud_instance, pycloudlib.EC2):
+        cloud_instance.upload_key(
+            public_key_path=cloud_instance.config["public_key_path"],
+            private_key_path=cloud_instance.config["private_key_path"],
+            name=cloud_instance.tag,
+        )
+
+    yield cloud_instance
+
+    if isinstance(cloud_instance, pycloudlib.EC2):
+        cloud_instance.delete_key(name=cloud_instance.tag)
+
+
+@contextmanager
+def launch_instance(
+    cloud_instance: BaseCloud, **kwargs
+) -> Generator[BaseInstance, None, None]:
+    instance = cloud_instance.launch(**kwargs)
+    try:
+        yield instance
+    finally:
+        instance.delete()
+
+
+def assert_example_output(instance: BaseInstance):
+    example_output = instance.execute("cat /var/tmp/example.txt").stdout
+    assert example_output == "hello"
+
+
+def exercise_push_pull(instance: BaseInstance):
+    with TemporaryDirectory() as tmpdir:
+        push_path = Path(tmpdir).joinpath("pushed")
+        push_path.write_text("pushed", encoding="utf-8")
+        instance.push_file(str(push_path), "/var/tmp/pushed")
+        assert (
+            "pushed" == instance.execute("cat /var/tmp/pushed").stdout.strip()
+        )
+
+        instance.execute("echo 'pulled' > /var/tmp/pulled")
+        pull_path = Path(tmpdir).joinpath("pulled")
+        instance.pull_file("/var/tmp/pulled", str(pull_path))
+        assert pull_path.read_text(encoding="utf-8").strip() == "pulled"
+
+
+def exercise_instance(instance: BaseInstance):
+    with suppress(NotImplementedError):
+        assert instance.name is not None
+
+    with suppress(NotImplementedError):
+        assert ipaddress.ip_address(instance.ip)
+
+    with suppress(NotImplementedError):
+        ip_address = instance.add_network_interface()
+        try:
+            assert ipaddress.ip_address(ip_address)
+        finally:
+            instance.remove_network_interface(ip_address)
+
+    exercise_push_pull(instance)
+
+    assert_example_output(instance)
+    instance.execute("sync")
+    boot_id = instance.get_boot_id()
+    instance.restart()
+    assert boot_id != instance.get_boot_id()
+
+    assert_example_output(instance)
```

Comment:
What are we testing here? Was the intent to validate that cloud-init didn't run twice and append to /var/tmp/example.txt?